JPEGImageCompression/image_compression.py

```diff
@@ -203,22 +210,93 @@ def get_yuv_image_from_rgb(self, pixel_array):
         img_yuv = cv2.cvtColor(rgb_img, cv2.COLOR_BGR2YUV)
         y, u, v = cv2.split(img_yuv)
 
-        lut_u, lut_v = make_lut_u(), make_lut_v()
+        if not mapped:
+            return y, u, v
+        else:
+            lut_u, lut_v = make_lut_u(), make_lut_v()
+
+            # Convert back to BGR so we display the images
+            y = cv2.cvtColor(y, cv2.COLOR_GRAY2BGR)
+            u = cv2.cvtColor(u, cv2.COLOR_GRAY2BGR)
+            v = cv2.cvtColor(v, cv2.COLOR_GRAY2BGR)
+
+            u_mapped = cv2.LUT(u, lut_u)
+            v_mapped = cv2.LUT(v, lut_v)
+
+            # Flip channels to RGB
+            y_rgb = y[:, :, [2, 1, 0]]
+            u_mapped_rgb = u_mapped[:, :, [2, 1, 0]]
+            v_mapped_rgb = v_mapped[:, :, [2, 1, 0]]
+
+            return y_rgb, u_mapped_rgb, v_mapped_rgb
+
+    def chroma_subsample_image(self, pixel_array, mode="4:2:2"):
+        """
+        Applies chroma subsampling to the image. Modes supported are the most common ones: 4:2:2 and 4:2:0.
+
+        @param: pixel_array - the image to be processed
+        @param: mode - a string, either `4:2:2` or `4:2:0`, corresponding to 4:2:2 and 4:2:0 subsampling respectively.
+        @param: image - returns back the image in RGB format with subsampling applied
+        """
+        assert mode in (
+            "4:2:2",
+            "4:2:0",
+        ), "Please choose one of the following {'4:2:2', '4:2:0'}"
+
+        y, u, v = self.get_yuv_image_from_rgb(pixel_array, mapped=False)
+
+        out_u = u.copy()
+        out_v = v.copy()
+        # Downsample with a window of 2 in the horizontal direction
+        if mode == "4:2:2":
+            # first the u channel
+            for i in range(0, u.shape[0], 2):
+                out_u[i : i + 2] = np.mean(u[i : i + 2], axis=0)
+
+            # then the v channel
+            for i in range(0, v.shape[0], 2):
+                out_v[i : i + 2] = np.mean(v[i : i + 2], axis=0)
+
+        # Downsample with a window of 2 in both directions
+        elif mode == "4:2:0":
+            for i in range(0, u.shape[0], 2):
+                for j in range(0, u.shape[1], 2):
+                    out_u[i : i + 2, j : j + 2] = np.mean(u[i : i + 2, j : j + 2])
```

Comment:
Minor nit: stylistically, when slicing indices of a list/array, make sure there are no spaces in between any characters.

E.g here, it should be `out_u[i:i+2, j:j+2] = np.mean(u[i:i+2, j:j+2])` -- this seems to be the accepted standard for style guidelines of slicing -- https://stackoverflow.com/questions/8092513/style-formatting-the-slice-operator

Reply:
Gotcha!

Reply:
Actually, looking a bit more into it, seems like black (my formatter) has already had some issues that have been resolved regarding this topic.

Acording to [PEP-8](https://www.python.org/dev/peps/pep-0008/#:~:text=%23%20Correct%3A%0Aham%5B1%3A9%5D%2C%20ham%5B1%3A9%3A3%5D%2C%20ham%5B%3A9%3A3%5D%2C%20ham%5B1%3A%3A3%5D%2C%20ham%5B1%3A9%3A%5D%0Aham%5Blower%3Aupper%5D%2C%20ham%5Blower%3Aupper%3A%5D%2C%20ham%5Blower%3A%3Astep%5D%0Aham%5Blower%2Boffset%20%3A%20upper%2Boffset%5D%0Aham%5B%3A%20upper_fn(x)%20%3A%20step_fn(x)%5D%2C%20ham%5B%3A%3A%20step_fn(x)%5D%0Aham%5Blower%20%2B%20offset%20%3A%20upper%20%2B%20offset%5D), the formatting I used is indeed compliant with the style guides.